crates/runtime/src/ch/blocks.rs

```diff
@@ -146,6 +147,133 @@ impl Default for Block {
     }
 }
 
+// TODO FIXME: The best approach is to unify the server and client block definitions
+// to avoid conversions
+#[inline]
+fn value_ref_to_bytes<'a>(value_ref: &'a ValueRefEnum<'a>) -> &'a [u8] {
+    match value_ref {
+        ValueRefEnum::String(bytes) => bytes,
+        ValueRefEnum::UInt8(u) => u.as_bytes(),
+        ValueRefEnum::UInt16(u) => u.as_bytes(),
+        ValueRefEnum::UInt32(u) => u.as_bytes(),
+        ValueRefEnum::UInt64(u) => u.as_bytes(),
+        ValueRefEnum::Int8(i) => i.as_bytes(),
+        ValueRefEnum::Int16(i) => i.as_bytes(),
+        ValueRefEnum::Int32(i) => i.as_bytes(),
+        ValueRefEnum::Int64(i) => i.as_bytes(),
+        ValueRefEnum::Date(d) => &d.0,
+        ValueRefEnum::DateTime(dt) => &dt.0,
+        ValueRefEnum::DateTime64(dt64) => dt64.0.as_bytes(),
+        ValueRefEnum::Array8(_)
+        | ValueRefEnum::Array16(_)
+        | ValueRefEnum::Array32(_)
+        | ValueRefEnum::Array64(_)
+        | ValueRefEnum::Array128(_)
+        | ValueRefEnum::Float32(_)
+        | ValueRefEnum::Float64(_)
+        | ValueRefEnum::Enum(_)
+        | ValueRefEnum::Ip4(_)
+        | ValueRefEnum::Ip6(_)
+        | ValueRefEnum::Uuid(_)
+        | ValueRefEnum::Decimal32(_)
+        | ValueRefEnum::Decimal64(_) => unimplemented!(),
+    }
+}
+
+// TODO FIXME: The best approach is to unify the server and client block definitions
+// to avoid conversions
+#[inline]
+fn sqltype_to_bqltype(sqltype: SqlType) -> BqlType {
+    match sqltype {
+        SqlType::UInt8 => BqlType::UInt(8),
+        SqlType::UInt16 => BqlType::UInt(16),
+        SqlType::UInt32 => BqlType::UInt(32),
+        SqlType::UInt64 => BqlType::UInt(64),
+        SqlType::Int8 => BqlType::Int(8),
+        SqlType::Int16 => BqlType::Int(16),
+        SqlType::Int32 => BqlType::Int(32),
+        SqlType::Int64 => BqlType::Int(64),
+        SqlType::String => BqlType::String,
+        SqlType::FixedString(i) => BqlType::FixedString(i as u8),
+        SqlType::Float32 => BqlType::Float(32),
+        SqlType::Float64 => BqlType::Float(64),
+        SqlType::Date => BqlType::Date,
+        SqlType::DateTime(tz) => match tz {
+            Some(tz) => {
+                BqlType::DateTimeTz(TimeZoneId(TimeZoneId::calc_offset_of_tz(tz)))
+            }
+            None => BqlType::DateTime,
+        },
+        SqlType::DateTime64(id, tz) => {
+            BqlType::DateTimeTz(TimeZoneId(TimeZoneId::calc_offset_of_tz(tz)))
+        }
+        SqlType::Decimal(x, y) => BqlType::Decimal(x, y),
+        SqlType::LowCardinality => BqlType::LowCardinalityTinyText,
+        SqlType::Ipv4
+        | SqlType::Ipv6
+        | SqlType::Uuid
+        | SqlType::Enum8
+        | SqlType::Enum16
+        | SqlType::Array => unimplemented!(),
+    }
+}
+
+// TODO FIXME: The best approach is to unify the server and client block definitions
+// to avoid conversions
+impl From<ServerBlock> for Block {
```

Comment:
from the algo, we use way: (client)ServerBlock (columnar) -> row-by-row -> (TB)Block (columnar). We should expose some methods via ServerBlock to directly convert the internal data of ServerBlock  to Block. And if this is ok, the rewriting for string should be not necessary as well.

Reply:
Yeah. My rough idea is that the next step is to directly unify the client and server block definitions, so this is a simple implementation without any abstraction, but the refactoring scope should be relatively large, requiring modification of the client code, and can be implemented by creating a separate branch. But this is the priority of the current tasks.